fdbclient/NativeAPI.actor.cpp

```diff
@@ -5929,8 +6020,35 @@ ACTOR Future<Version> extractReadVersion(Reference<TransactionState> trState,
 	return rep.version;
 }
 
+bool rkThrottlingCooledDown(DatabaseContext* cx) {
+	if (cx->lastTimedRkThrottle == 0.0) {
+		return true;
+	}
+	return (now() - cx->lastTimedRkThrottle > CLIENT_KNOBS->GRV_CACHE_RK_COOLDOWN);
+}
+
 Future<Version> Transaction::getReadVersion(uint32_t flags) {
 	if (!readVersion.isValid()) {
+		if (!CLIENT_KNOBS->FORCE_GRV_CACHE_OFF && !trState->options.skipGrvCache &&
+		    (deterministicRandom()->random01() <= CLIENT_KNOBS->DEBUG_USE_GRV_CACHE_CHANCE ||
+		     trState->options.useGrvCache) &&
+		    rkThrottlingCooledDown(getDatabase().getPtr())) {
+			// Upon our first request to use cached RVs, start the background updater
+			if (!trState->cx->grvUpdateHandler.isValid()) {
+				trState->cx->grvUpdateHandler = backgroundGrvUpdater(getDatabase().getPtr());
+			}
+			Version rv = trState->cx->getCachedRV();
+			double lastTime = trState->cx->getLastTimedGRV();
+			double requestTime = now();
+			if (requestTime - lastTime <= CLIENT_KNOBS->MAX_VERSION_CACHE_LAG && rv != Version(0)) {
+				ASSERT(!debug_checkVersionTime(rv, requestTime, "CheckStaleness"));
+				readVersion = rv;
+				return readVersion;
+			} // else go through regular GRV path
+		}
+		if (CLIENT_KNOBS->FORCE_GRV_CACHE_OFF && trState->cx->grvUpdateHandler.isValid()) {
```

Comment:
It can't be set, but I think that would not account for the scenario where the knob is toggled while running. I think that's a scenario we need to check for with dynamic knobs(?) coming

Reply:
Ok, I see. In that case, I think I like Vaidas's suggestion.